File: tools/scrapers/centrometal-scraper.js
/**
 * Centrometal Heater Data Scraper (Production)
 *
 * Scrapes real-time heater data and saves to PostgreSQL database
 *
 * Usage:
 *   node tools/scrapers/centrometal-scraper.js
 */

import { chromium } from 'playwright';
import dotenv from 'dotenv';
import { saveHeaterStatus, saveHeaterStatistics } from '../../backend/src/services/heaterService.js';

dotenv.config();

// Configuration
const PORTAL_URL = process.env.CENTROMETAL_PORTAL_URL || 'https://www.portal.centrometal.hr';
const LOGIN_URL = `${PORTAL_URL}/login`;
const DASHBOARD_URL = `${PORTAL_URL}/#/widgetsgrid`;
const INSTALLATION_ID = process.env.CENTROMETAL_INSTALLATION_ID || '56F0C163';
const DATABASE_ID = 4210; // Discovered from debug script

async function scrapeHeaterData() {
  console.log('[*] Starting Centrometal heater scraper...\n');

  // Validate credentials
  if (!process.env.CENTROMETAL_USERNAME || !process.env.CENTROMETAL_PASSWORD) {
    console.error('[ERR] Missing credentials in .env file');
    process.exit(1);
  }

  const browser = await chromium.launch({
    headless: true  // Run in background
  });

  const context = await browser.newContext();
  const page = await context.newPage();

  // Storage for captured data
  let temperatureData = null;
  let errorData = null;
  let statusData = null;

  // Intercept API responses to capture heater data
  page.on('response', async response => {
    const url = response.url();

    // Optionally log all API calls for debugging
    // if (url.includes('centrometal.hr') && (url.includes('/api/') || url.includes('/wdata/'))) {
    //   console.log(`[DEBUG] API call: ${url.split('?')[0]}`);
    // }

    try {
      // Capture temperature data from traffic endpoint
      if (url.includes('/wdata/data/multi/traffic/')) {
        const data = await response.json();

        if (data.rows && data.rows.length > 0) {
          const latest = data.rows[data.rows.length - 1];
          // Columns: timestamp, boiler_temp, flue_gas_temp, return_temp, dhw_temp
          temperatureData = {
            timestamp: latest[0],
            boilerTemp: parseFloat(latest[1]) || null,
            flueGasTemp: parseFloat(latest[2]) || null,
            returnTemp: parseFloat(latest[3]) || null,
            dhwTemp: parseFloat(latest[4]) || null
          };
          console.log('[OK] Captured temperature data:', temperatureData);
        }
      }

      // Capture cumulative statistics and status data
      if (url.includes('/wdata/data/installation-status-all')) {
        const data = await response.json();

        if (data && data[DATABASE_ID] && data[DATABASE_ID].params) {
          const params = data[DATABASE_ID].params;

          statusData = {
            // Heater state
            state: params.B_STATE?.v || null,
            cmd: params.B_CMD?.v || null,

            // Current measurements
            mainTemp: parseFloat(params.B_Tk1?.v) || null,
            supplyTemp: parseFloat(params.B_Tdpl1?.v) || null,
            returnTemp: parseFloat(params.B_Tpov1?.v) || null,
            dhwTemp: parseFloat(params.B_Tptv1?.v) || null,
            externalTemp: parseFloat(params.B_Out1?.v) || null,
            oxygenLevel: parseFloat(params.B_Oxy1?.v) || null,
            fanSpeed: parseFloat(params.B_fan?.v) || null,

            // Pump statuses
            pumpP1Status: params.B_P1?.v || null,
            pumpP2Status: params.B_P2?.v || null,
            pumpP3Status: params.B_P3?.v || null,
            pumpP1Demand: params.B_zahP1?.v || null,
            pumpP2Demand: params.B_zahP2?.v || null,
            pumpP3Demand: params.B_zahP3?.v || null,

            // Fan status
            fanStatus: params.B_FotV?.v || null,

            // Cumulative statistics
            burnerWorkMinutes: parseInt(params.CNT_0?.v) || null,
            burnerStartCount: parseInt(params.CNT_1?.v) || null,
            fuelConsumedKg: parseFloat(params.CNT_2?.v) || null,
            flameDurationMinutes: parseInt(params.CNT_3?.v) || null,
            fanWorkingMinutes: parseInt(params.CNT_4?.v) || null,
            pelletAccumulatorMinutes: parseInt(params.CNT_5?.v) || null,
            vacuumBoilerMinutes: parseInt(params.CNT_6?.v) || null,
            vacuumTurbineMinutes: parseInt(params.CNT_7?.v) || null,
            vacuumTurbineCycles: parseInt(params.CNT_8?.v) || null,
            timeOnState01: parseInt(params.CNT_8?.v) || null,
            timeOnState02: parseInt(params.CNT_9?.v) || null,
          };

          console.log('[OK] Captured status and statistics data');
          console.log('[OK] Heater state:', statusData.state);
          console.log('[OK] Burner work: ', statusData.burnerWorkMinutes, 'min');
          console.log('[OK] Fuel consumed:', statusData.fuelConsumedKg, 'kg');
        }
      }

      // Capture error list
      if (url.includes('/wdata/data/multi/errors-list')) {
        const data = await response.json();
        if (data.rows && data.rows.length > 0) {
          errorData = data;
          console.log(`[WARN] Found ${data.rows.length} error events`);
        }
      }

    } catch (e) {
      // Ignore non-JSON responses
    }
  });

  try {
    // Step 1: Login
    console.log('[*] Logging in...');
    await page.goto(LOGIN_URL, { waitUntil: 'domcontentloaded', timeout: 30000 });
    await page.waitForTimeout(2000);

    // Find and fill login form
    const usernameSelectors = [
      'input[name="username"]',
      'input[name="_username"]',
      'input[type="email"]',
      'input[type="text"]'
    ];

    for (const selector of usernameSelectors) {
      try {
        const field = await page.$(selector);
        if (field && await field.isVisible()) {
          await field.fill(process.env.CENTROMETAL_USERNAME);
          console.log(`[OK] Filled username`);
          break;
        }
      } catch (e) { }
    }

    const passwordField = await page.$('input[type="password"]');
    if (passwordField) {
      await passwordField.fill(process.env.CENTROMETAL_PASSWORD);
      console.log('[OK] Filled password');
    }

    // Submit
    const submitButton = await page.$('button[type="submit"]');
    if (submitButton) {
      await submitButton.click();
      console.log('[OK] Submitted login form');

      // Wait for navigation
      try {
        await page.waitForNavigation({ waitUntil: 'domcontentloaded', timeout: 10000 });
      } catch (e) {
        // Might already be on dashboard
      }
    }

    // Step 2: Navigate to dashboard
    console.log('[*] Loading dashboard...');
    await page.goto(DASHBOARD_URL, { waitUntil: 'domcontentloaded', timeout: 30000 });

    // Wait for data to load
    console.log('[*] Waiting for dashboard data...');
    await page.waitForTimeout(15000); // Give time for API calls to complete

    // Step 3: Save to database
    if (statusData || temperatureData) {
      console.log('[*] Saving to database...');

      // Helper to convert string/number to boolean
      const toBoolean = (val) => {
        if (val === null || val === undefined) return null;
        if (typeof val === 'boolean') return val;
        if (typeof val === 'string') {
          const num = parseInt(val);
          return num === 1;
        }
        if (typeof val === 'number') {
          return val === 1;
        }
        return null;
      };

      // Save heater status (temperatures, state, component statuses)
      const heaterStatus = {
        state: statusData?.state || 'UNKNOWN',
        isLive: true,
        mainTemperature: statusData?.mainTemp || temperatureData?.boilerTemp || null,
        supplyTemperature: statusData?.supplyTemp || temperatureData?.flueGasTemp || null,
        returnTemperature: statusData?.returnTemp || temperatureData?.returnTemp || null,
        bufferTankTemperature: null,
        radiatorTemperature: null,
        externalTemperature: statusData?.externalTemp || null,
        dhwTemperature: statusData?.dhwTemp || temperatureData?.dhwTemp || null,
        fuelLevelStatus: null,
        oxygenLevel: statusData?.oxygenLevel || null,
        pumpP1Status: toBoolean(statusData?.pumpP1Status),
        pumpP2Status: toBoolean(statusData?.pumpP2Status),
        pumpP3Status: toBoolean(statusData?.pumpP3Status),
        pumpP1Demand: toBoolean(statusData?.pumpP1Demand),
        pumpP2Demand: toBoolean(statusData?.pumpP2Demand),
        pumpP3Demand: toBoolean(statusData?.pumpP3Demand),
        fanStatus: toBoolean(statusData?.fanStatus),
        configurationType: DATABASE_ID,
        installationId: INSTALLATION_ID
      };

      const statusResult = await saveHeaterStatus(heaterStatus);
      console.log(`[OK] Saved status to database (ID: ${statusResult.id})`);

      // Save heater statistics if available
      if (statusData) {
        const heaterStats = {
          timestamp: new Date(),
          burnerWorkMinutes: statusData.burnerWorkMinutes,
          burnerStartCount: statusData.burnerStartCount,
          fuelConsumedKg: statusData.fuelConsumedKg,
          pelletAccumulatorMinutes: statusData.pelletAccumulatorMinutes,
          fanWorkingMinutes: statusData.fanWorkingMinutes,
          vacuumBoilerMinutes: statusData.vacuumBoilerMinutes,
          vacuumTurbineMinutes: statusData.vacuumTurbineMinutes,
          vacuumTurbineCycles: statusData.vacuumTurbineCycles,
          timeOnState01: statusData.timeOnState01,
          timeOnState02: statusData.timeOnState02,
          timeOnState03: null,
          timeOnState04: null,
          timeOnState05: null,
          timeOnState06: null,
        };

        const statsResult = await saveHeaterStatistics(heaterStats);
        console.log('[OK] Saved statistics to database');
        console.log('[OK] Burner work:', statusData.burnerWorkMinutes, 'min');
        console.log('[OK] Fuel consumed:', statusData.fuelConsumedKg, 'kg');
        console.log('[OK] Burner starts:', statusData.burnerStartCount);
      }

    } else {
      console.log('[WARN] No data captured');
    }

    console.log('\n[OK] Scraping completed successfully!');

  } catch (error) {
    console.error('[ERR] Scraping failed:', error.message);
    throw error;

  } finally {
    await browser.close();
  }
}

// Run
scrapeHeaterData()
  .then(() => process.exit(0))
  .catch(error => {
    console.error('[ERR] Fatal error:', error);
    process.exit(1);
  });

File: tools/scrapers/centrometal_scraper.py
"""
Centrometal PelTec Lambda Heater Data Scraper

A lightweight Python scraper that fetches real-time heater data from the Centrometal portal
and saves it to the PostgreSQL database.

Usage:
    python tools/scrapers/centrometal_scraper.py

Environment Variables Required:
    CENTROMETAL_USERNAME
    CENTROMETAL_PASSWORD
    CENTROMETAL_INSTALLATION_ID
    DATABASE_URL (PostgreSQL connection string)
"""

import os
import sys
import requests
import json
from datetime import datetime
from dotenv import load_dotenv
import psycopg

# Load environment variables
load_dotenv()

# Configuration
PORTAL_URL = os.getenv('CENTROMETAL_PORTAL_URL', 'https://www.portal.centrometal.hr')
LOGIN_URL = f"{PORTAL_URL}/login_check"
INSTALLATION_ID = os.getenv('CENTROMETAL_INSTALLATION_ID', '56F0C163')
DATABASE_URL = os.getenv('DATABASE_URL')

# Session management
session = requests.Session()
session.headers.update({
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
    'Accept': 'application/json, text/plain, */*',
    'Content-Type': 'application/json;charset=UTF-8'
})


class CentrometalScraper:
    """Scraper for Centrometal heater data"""

    def __init__(self):
        self.username = os.getenv('CENTROMETAL_USERNAME')
        self.password = os.getenv('CENTROMETAL_PASSWORD')
        self.installation_db_id = None  # Will be discovered
        self.logged_in = False

        if not self.username or not self.password:
            raise ValueError("Missing CENTROMETAL_USERNAME or CENTROMETAL_PASSWORD in .env")

    def login(self):
        """Login to Centrometal portal and establish session"""
        print('[*] Logging in to Centrometal portal...')

        try:
            # Form data for login (discovered from debug script)
            login_data = {
                '_username': self.username,
                '_password': self.password,
                '_remember_me': 'on'
            }

            response = session.post(
                LOGIN_URL,
                data=login_data,  # Use form data, not JSON
                headers={
                    'Content-Type': 'application/x-www-form-urlencoded',
                    'Referer': f'{PORTAL_URL}/login'
                },
                allow_redirects=True
            )

            # Check if login was successful by looking for session cookie
            if 'PHPSESSID' in session.cookies:
                self.logged_in = True
                print(f'[OK] Login successful! Session: {session.cookies.get("PHPSESSID")[:10]}...')
                return True
            else:
                print(f'[ERR] Login failed - no session cookie. Status: {response.status_code}')
                return False

        except Exception as e:
            print(f'[ERR] Login error: {e}')
            return False

    def load_dashboard(self):
        """Load the dashboard page to establish session state"""
        print('[*] Loading dashboard to establish session...')

        try:
            dashboard_url = f'{PORTAL_URL}/#/widgetsgrid'
            response = session.get(dashboard_url, timeout=10)

            if response.status_code == 200:
                print('[OK] Dashboard loaded successfully')
                return True
            else:
                print(f'[WARN] Dashboard returned status {response.status_code}')
                return False

        except Exception as e:
            print(f'[WARN] Error loading dashboard: {e}')
            return False

    def discover_installation_db_id(self):
        """Discover the database ID for the installation"""
        print(f'[*] Discovering database ID for installation {INSTALLATION_ID}...')

        try:
            # Try to get installation autocomplete data
            response = session.post(
                f'{PORTAL_URL}/data/autocomplete/installation',
                json={},
                timeout=10
            )

            if response.status_code == 200:
                data = response.json()
                # Search for our installation ID in the response
                for item in data:
                    if item.get('serial') == INSTALLATION_ID:
                        self.installation_db_id = item.get('id')
                        print(f'[OK] Found database ID: {self.installation_db_id}')
                        return self.installation_db_id

            # Fallback: Use hardcoded value discovered from debug
            self.installation_db_id = 4210
            print(f'[WARN]  Using fallback database ID: {self.installation_db_id}')
            return self.installation_db_id

        except Exception as e:
            print(f'[WARN]  Error discovering ID: {e}, using fallback')
            self.installation_db_id = 4210
            return self.installation_db_id

    def fetch_temperature_data(self):
        """Fetch real-time temperature and status data"""
        print('[*]  Fetching temperature data...')

        if not self.installation_db_id:
            self.discover_installation_db_id()

        try:
            response = session.post(
                f'{PORTAL_URL}/api/inst/control/multiple',
                json={},
                headers={
                    'Accept': 'application/json, text/plain, */*',
                    'Content-Type': 'application/json;charset=UTF-8',
                    'Referer': f'{PORTAL_URL}/',
                },
                timeout=10
            )

            if response.status_code == 200:
                print(f'[DEBUG] Response content: {response.text[:200]}')
                if not response.text:
                    print('[WARN] Empty response body')
                    return None

                data = response.json()

                # Parse the Google Charts format returned
                if 'rows' in data and len(data['rows']) > 0:
                    latest_row = data['rows'][-1]  # Get most recent data

                    # Columns are: timestamp, boiler_temp, flue_gas_temp, return_temp, dhw_temp
                    return {
                        'timestamp': latest_row[0],
                        'boiler_temperature': float(latest_row[1]) if latest_row[1] else None,
                        'flue_gas_temperature': float(latest_row[2]) if latest_row[2] else None,
                        'return_temperature': float(latest_row[3]) if latest_row[3] else None,
                        'dhw_temperature': float(latest_row[4]) if latest_row[4] else None,
                    }

            print(f'[WARN]  Temperature fetch returned status {response.status_code}')
            return None

        except Exception as e:
            print(f'[ERR] Error fetching temperature data: {e}')
            return None

    def fetch_parameter_list(self):
        """Fetch detailed parameter list including all sensors"""
        print('[*] Fetching parameter list...')

        try:
            response = session.get(
                f'{PORTAL_URL}/wdata/data/parameter-list/{INSTALLATION_ID}',
                headers={
                    'Accept': 'application/json, text/plain, */*',
                    'Referer': f'{PORTAL_URL}/',
                },
                timeout=10
            )

            if response.status_code == 200:
                data = response.json()
                print(f'[OK] Retrieved {len(data.get("parameters", []))} parameter groups')
                return data

            print(f'[WARN]  Parameter list fetch returned status {response.status_code}')
            return None

        except Exception as e:
            print(f'[ERR] Error fetching parameters: {e}')
            return None

    def fetch_errors(self):
        """Fetch error/event list"""
        print('[WARN]  Fetching error list...')

        if not self.installation_db_id:
            self.discover_installation_db_id()

        try:
            response = session.post(
                f'{PORTAL_URL}/wdata/data/multi/errors-list/{self.installation_db_id}',
                json={},
                headers={
                    'Accept': 'application/json, text/plain, */*',
                    'Content-Type': 'application/json;charset=UTF-8',
                    'Referer': f'{PORTAL_URL}/',
                },
                timeout=10
            )

            if response.status_code == 200:
                data = response.json()
                if 'rows' in data and len(data['rows']) > 0:
                    print(f'[OK] Found {len(data["rows"])} events')
                    return data
                else:
                    print('[OK] No errors found (good!)')
                    return None

            return None

        except Exception as e:
            print(f'[ERR] Error fetching errors: {e}')
            return None

    def save_to_database(self, heater_data):
        """Save heater data to PostgreSQL database"""
        print('[*] Saving to database...')

        if not DATABASE_URL:
            print('[WARN]  DATABASE_URL not set, skipping database save')
            return False

        try:
            with psycopg.connect(DATABASE_URL) as conn:
                with conn.cursor() as cursor:
                    # Insert heater status
                    query = """
                        INSERT INTO heater_status (
                            state, is_live,
                            main_temperature, supply_temperature, return_temperature,
                            dhw_temperature,
                            configuration_type, installation_id
                        ) VALUES (
                            %s, %s, %s, %s, %s, %s, %s, %s
                        ) RETURNING id
                    """

                    values = (
                        heater_data.get('state', 'UNKNOWN'),
                        True,  # is_live
                        heater_data.get('boiler_temperature'),
                        heater_data.get('flue_gas_temperature'),
                        heater_data.get('return_temperature'),
                        heater_data.get('dhw_temperature'),
                        heater_data.get('database_id'),  # configuration_type (INTEGER)
                        INSTALLATION_ID  # installation_id (string)
                    )

                    cursor.execute(query, values)
                    record_id = cursor.fetchone()[0]

                    conn.commit()

            print(f'[OK] Saved to database (ID: {record_id})')
            return True

        except Exception as e:
            print(f'[ERR] Database error: {e}')
            return False

    def scrape(self):
        """Main scraping function"""
        print('[*] Starting Centrometal heater scraper...\n')

        # Step 1: Login
        if not self.login():
            print('[ERR] Scraping failed - login unsuccessful')
            return None

        # Step 2: Load dashboard (required for API access)
        self.load_dashboard()

        # Step 3: Discover installation database ID
        self.discover_installation_db_id()

        # Step 4: Fetch data
        temp_data = self.fetch_temperature_data()
        param_data = self.fetch_parameter_list()
        error_data = self.fetch_errors()

        # Combine data
        heater_data = {
            'timestamp': datetime.now().isoformat(),
            'installation_id': INSTALLATION_ID,
            'database_id': self.installation_db_id,
        }

        if temp_data:
            heater_data.update(temp_data)

        # Step 5: Save to database
        if heater_data:
            self.save_to_database(heater_data)

            print('\n[OK] Scraping completed successfully!')
            print(f'[*] Data: {json.dumps(heater_data, indent=2, default=str)}')
            return heater_data
        else:
            print('[ERR] No data retrieved')
            return None


def main():
    """Main entry point"""
    try:
        scraper = CentrometalScraper()
        data = scraper.scrape()

        if data:
            sys.exit(0)  # Success
        else:
            sys.exit(1)  # Failure

    except Exception as e:
        print(f'[ERR] Fatal error: {e}')
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == '__main__':
    main()
